vsintegration/src/FSharp.Editor/FsiCommandService.fs

```diff
@@ -0,0 +1,92 @@
+        | VSConstants.S_OK, pkg ->
+            pkg :?> Package
+        | _ -> null)
+
+    interface IMenuCommand with
+        member val IsAdded = false with get, set
+        member val NextTarget = null with get, set
+
+        member x.Exec (pguidCmdGroup, nCmdId, nCmdexecopt, pvaIn, pvaOut) =
+            if pguidCmdGroup = VSConstants.VsStd11 then
+                if nCmdId = uint32 VSConstants.VSStd11CmdID.ExecuteSelectionInInteractive then
+                    Hooks.OnMLSend projectSystemPackage.Value FsiEditorSendAction.ExecuteSelection null null
+                elif nCmdId = uint32 VSConstants.VSStd11CmdID.ExecuteLineInInteractive then
+                    Hooks.OnMLSend projectSystemPackage.Value FsiEditorSendAction.ExecuteLine null null
+                VSConstants.S_OK
+            elif pguidCmdGroup = Guids.guidInteractive then
+                if nCmdId = uint32 Guids.cmdIDDebugSelection then
+                    Hooks.OnMLSend projectSystemPackage.Value FsiEditorSendAction.DebugSelection null null
+                VSConstants.S_OK
+            else
+                let x = x :> IMenuCommand
+                x.NextTarget.Exec(&pguidCmdGroup, nCmdId, nCmdexecopt, pvaIn, pvaOut)
+
+        member x.QueryStatus (pguidCmdGroup, cCmds, prgCmds, pCmdText) =
+            if pguidCmdGroup = VSConstants.VsStd11 then
+                if prgCmds.[0].cmdID = uint32 VSConstants.VSStd11CmdID.ExecuteSelectionInInteractive then
+                    prgCmds.[0].cmdf <- uint32 (OLECMDF.OLECMDF_SUPPORTED ||| OLECMDF.OLECMDF_ENABLED)
+                elif prgCmds.[0].cmdID = uint32 VSConstants.VSStd11CmdID.ExecuteLineInInteractive then
+                    prgCmds.[0].cmdf <- uint32 (OLECMDF.OLECMDF_SUPPORTED ||| OLECMDF.OLECMDF_ENABLED ||| OLECMDF.OLECMDF_DEFHIDEONCTXTMENU)
+                VSConstants.S_OK
+            elif pguidCmdGroup = Guids.guidInteractive then
+                if prgCmds.[0].cmdID = uint32 Guids.cmdIDDebugSelection then
+                    let dbgState = Hooks.GetDebuggerState projectSystemPackage.Value
+                    if dbgState = FsiDebuggerState.AttachedNotToFSI then
+                        prgCmds.[0].cmdf <- uint32 OLECMDF.OLECMDF_INVISIBLE
+                    else
+                        prgCmds.[0].cmdf <- uint32 (OLECMDF.OLECMDF_SUPPORTED ||| OLECMDF.OLECMDF_ENABLED)
+                VSConstants.S_OK
+            else
+                let x = x :> IMenuCommand
+                x.NextTarget.QueryStatus(&pguidCmdGroup, cCmds, prgCmds, pCmdText)
+
+[<Export(typeof<IWpfTextViewCreationListener>)>]
+[<ContentType("F#")>]
+[<TextViewRole(PredefinedTextViewRoles.PrimaryDocument)>]
+type FsiCommandFilterProvider [<ImportingConstructor>] 
+    ([<Import(typeof<SVsServiceProvider>)>] serviceProvider: System.IServiceProvider,
+     editorFactory: IVsEditorAdaptersFactoryService) =
+  
+    let addCommandFilter (viewAdapter: IVsTextView) (commandFilter: IMenuCommand) =
+        if not commandFilter.IsAdded then
```

Comment:
Why do we need ``IsAdded``?  Looking at the logic the command filter has only just been freshly created in any case?